### PR TITLE
Create a new client based on Java 11 HttpClient

### PR DESCRIPTION
**Pull request by laeubi**
As discussed here: https://bugs.eclipse.org/bugs/show_bug.cgi?id=580896

**This is just a draft to allow sharing/discussing the progress and not yet ready to be merged or fully reviewed.**

### CODE REVIEW

**Comment by laeubi** [comment]
First question is there seems to be a `IHttpClientFactory` service but it is in an internal package, is this a requirement for a provider to have such thing?

**Comment by laeubi** [comment]
Same for `IHttpClientModifier` as far as I understand all this is so that users of ECF have a way to further adjust configuration of (vendor specific) connector implementations, Question would be if this should be replicated for a Java based client as well or if this is optional and could be omitted.

File: providers/bundles/org.eclipse.ecf.provider.filetransfer.httpclientjava/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+Bundle-Version: 1.0.300.qualifier
+Bundle-Vendor: %plugin.provider
+Bundle-Localization: plugin
+Automatic-Module-Name: org.eclipse.ecf.provider.filetransfer.httpclientjava
+Bundle-RequiredExecutionEnvironment: JavaSE-11
+Require-Bundle: org.eclipse.equinox.common;bundle-version="3.13.0",
+ org.eclipse.ecf;bundle-version="3.9.101",
+ org.eclipse.core.jobs;bundle-version="3.10.800"
+Import-Package: javax.net.ssl,
+ org.apache.hc.client5.http;version="5.0.2",
```
Comment (akurtakov) [inline providers/bundles/org.eclipse.ecf.provider.filetransfer.httpclientjava/META-INF/MANIFEST.MF:14]: This looks like copy paste error, right? Apache httpclient shouldn't be needed by this bundle.

Comment (laeubi) [reply providers/bundles/org.eclipse.ecf.provider.filetransfer.httpclientjava/META-INF/MANIFEST.MF:14]: Currently this is just a copy of the hc5 bundle with just package/project names adjusted.

Comment (akurtakov) [reply providers/bundles/org.eclipse.ecf.provider.filetransfer.httpclientjava/META-INF/MANIFEST.MF:14]: Ah, I thought it's implementation using jvm's httpclient.

Comment (laeubi) [reply providers/bundles/org.eclipse.ecf.provider.filetransfer.httpclientjava/META-INF/MANIFEST.MF:14]: Yes that's the goal, but currently it has no changed implementation and I'm using this as a base to investigate what needs changed.

**Comment by scottslewis** [comment]
> Same for `IHttpClientModifier` as far as I understand all this is so that users of ECF have a way to further adjust configuration of (vendor specific) connector implementations, Question would be if this should be replicated for a Java based client as well or if this is optional and could be omitted.

The httpclient5 provider was implemented by Yatta Systems...can't remember the name of the committer at the moment.   As for IHttpClientModifier and IHttpClientFactory...yes it's strange that they made these (internal by package name) into services...my guess is that they intended to add some sort of extension capability for httpclient5 provider using these services but perhaps never got to it.  In any event afaik they are not used by any other bundle/fragment at the moment.

**Comment by laeubi** [comment]
@akurtakov @scottslewis I have now a first version that **compiles**, I wonder if I can enable some kind of integration test that uses the implementation to check its **working** as expected?

**Comment by akurtakov** [comment]
IIRC @scottslewis had some way of testing providers for compatibility.

**Comment by laeubi** [comment]
> IIRC @scottslewis had some way of testing providers for compatibility.

Alright, maybe it makes sense that some ECF committer takes over here? 
The code compiles and shows how to use the Java 11 API and I tried to make it as close as possible to the existing commons4/5 implementation but my knowledge of ECF is rather limited here to be more helpful :-)

**Comment by scottslewis** [comment]
Here is an httpclient5 test bundle here:   https://github.com/eclipse/ecf/tree/master/tests/bundles/org.eclipse.ecf.tests.filetransfer.httpclient5

The code in this is trivial/just test setup with the provider....most of the actual test code is reused in dependent test bundles (also in same directory).

A new test bundle should be created from this test httpclient5 bundle with references changed to new provider....just as was done with httpclient5 -> javahttpclient provider bundle itself.

Once the new test bundle is running I can add it to our tests feature and run as part of integration builds.

**Comment by scottslewis** [comment]
For reference here is our integration server:  https://ci.eclipse.org/ecf/

**Comment by scottslewis** [comment]
So that everyone is aware:

Because of the issue with signing, ECF had to produce another release (3.14.37) and this release now includes the Java 11 httpclient provider...both the bundle/plugin and the feature.

Here's the ECF 3.14.37 repo:

https://download.eclipse.org/rt/ecf/3.14.37/site.p2/3.14.37.v20221119-2038/

Both the Apache httpclient and the Java 11+ httpclient are present in the above repo.  The Apache httpclient (5.1.3/5.1.4) is being consumed by the platform for 4.26 (2022-12) release scheduled for this week.   

 The Java 11 httpclient provider can be installed into Eclipse (both before and after 2022-12 is released) and tested.   Note that if it's installed it has higher priority than the Apache httpclient provider and so will be preferred at download time.

**Comment by laeubi** [comment]
@scottslewis many thanks this sounds very useful!

**Pull request merged**